genienlp/data_utils/bootleg.py

```diff
@@ -198,35 +196,35 @@ def add_type_tokens(self, sentence, features):
         assert len(sentence_tokens) == len(features)
         sentence_plus_types_tokens = []
         i = 0
-        if self.args.add_types_to_text == 'insert' or self.args.add_qids_to_text == 'insert':
+        if self.args.add_entities_to_text == 'insert':
             while i < len(sentence_tokens):
                 token = sentence_tokens[i]
                 feat = features[i]
                 # token is an entity
-                if any([val != self.args.ned_features_default_val[0] for val in feat.type_id]):
+                if any([val != 0 for val in feat.type_id]):
                     final_token = '<e> '
                     final_types = ''
-                    if self.args.add_types_to_text != 'no':
+                    if 'type_id' in self.args.entity_attributes:
                         if self.bootleg_post_process_types:
                             all_types = ' | '.join(
-                                set(
-                                    self.qid2almond_type[self.id2typeqid[id]]
+                                sorted(
+                                    self.entityqid_to_type_vocab[self.id2typeqid[id]]
                                     for id in feat.type_id
-                                    if self.id2typeqid[id] in self.qid2almond_type
+                                    if self.id2typeqid[id] in self.entityqid_to_type_vocab
                                 )
                             )
                         else:
                             all_types = ' | '.join(
-                                set(
-                                    self.wikidataqid_to_type_vocab[self.id2typeqid[id]]
+                                sorted(
+                                    self.entityqid_to_type_vocab[self.id2typeqid[id]]
                                     for id in feat.type_id
                                     if id != self.args.db_unk_id
                                 )
                             )
                         final_types = '( ' + all_types + ' ) '
                     final_qids = ''
-                    if self.args.add_qids_to_text != 'no':
-                        all_qids = ' | '.join(set('Q' + str(id) for id in feat.qid))
+                    if 'qid' in self.args.entity_attributes:
+                        all_qids = ' | '.join(sorted('Q' + str(id) for id in feat.qid if id != -1))
                         final_qids = '[' + all_qids + ']'
```

Comment:
I think you should extract all the code that computes the string to append (with types/qids) into a method that takes the Entity object. The same code is used in insert or append mode.